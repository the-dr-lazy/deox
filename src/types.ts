import { ActionCreator } from './create-action-creator'
import { Action, AnyAction } from './action'

type Primitive = undefined | null | boolean | string | number | Function

export interface DeepImmutableArray<T>
  extends ReadonlyArray<DeepImmutable<T>> {}
export interface DeepImmutableMap<K, V>
  extends ReadonlyMap<DeepImmutable<K>, DeepImmutable<V>> {}
export type DeepImmutableObject<T> = {
  readonly [K in keyof T]: DeepImmutable<T[K]>
}

export type Immutable<T> = T extends Primitive
  ? T
  : T extends Array<infer U>
  ? ReadonlyArray<U>
  : T extends Map<infer K, infer V>
  ? ReadonlyMap<K, V>
  : Readonly<T>

export type DeepImmutable<T> = T extends Primitive
  ? T
  : T extends Array<infer U>
  ? DeepImmutableArray<U>
  : T extends Map<infer K, infer V>
  ? DeepImmutableMap<K, V>
  : DeepImmutableObject<T>

export type Reducer<State, Actions> = (
  prevState: DeepImmutable<State>,
  action: Actions
) => DeepImmutable<State> | State

export type ActionType<
  T extends ActionCreator<AnyAction> | Reducer<any, Action<any>>
> = T extends ActionCreator<AnyAction>
  ? ReturnType<T>
  : (T extends Reducer<any, infer U> ? U : never)
