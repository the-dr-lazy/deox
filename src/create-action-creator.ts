import { createAction, Action, AnyAction } from './create-action'

export type ExactActionCreator<TType extends string, TCallable extends <_T>(...args: any[]) => Action<TType>> = TCallable & {
  type: TType extends AnyAction ? TType['type'] : TType
  toString(): TType extends AnyAction ? TType['type'] : TType
}

export type ActionCreator<T extends AnyAction | string> = {
  (...args: any[]): T extends string ? Action<T> : T
  type: T extends AnyAction ? T['type'] : T
  toString(): T extends AnyAction ? T['type'] : T
}

export type Executor<TType extends string, TCallable extends <_T>(...args: any[]) => Action<TType>> = (
  resolve: <Payload = undefined, Meta = undefined>(
    payload?: Payload,
    meta?: Meta
  ) => Action<TType, Payload, Meta>
) => TCallable;

/**
 * Flux standard action creator factory
 * @example
 * createActionCreator('[Todo] truncate');
 * @example
 * createActionCreator(
 *   '[Todo] fetch rejected',
 *   resolve => (error: Error) => resolve(error)
 * );
 * @example
 * createActionCreator(
 *   '[Todo] fetch rejected',
 *   resolve => (error: Error, meta?: { status: number }) => resolve(error, meta)
 * )
 * @example
 * createActionCreator(
 *   '[Todo] add',
 *   resolve => (name: string, completed = false) => resolve({ name, completed })
 * )
 * @example
 * createActionCreator(
 *   '[Todo] add',
 *   resolve => (name: string, completed = false) => resolve({ name, completed }, 'Meta data of all todos')
 * )
 *
 */
export function createActionCreator<TType extends string>(type: TType): ExactActionCreator<TType, () => Action<TType>>;
export function createActionCreator<
  TType extends string,
  TCallable extends <_T>(...args: any[]) => Action<TType>
>(type: TType, executor: Executor<TType, TCallable>): ExactActionCreator<TType, TCallable>;
export function createActionCreator<
  TType extends string,
  TCallable extends <_T>(...args: any[]) => Action<TType>
>(type: TType, executor: Executor<TType, TCallable> = resolve => (() => resolve()) as TCallable) {
  const callable = executor((payload, meta) =>
    createAction(type, payload!, meta!)
  )

  return Object.assign(callable, {
    type,
    toString() {
      return type
    },
  })
}
