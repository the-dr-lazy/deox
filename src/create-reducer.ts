import {
  createHandlerMap,
  CreateHandlerMap,
  HandlerMap,
  InferActionFromHandlerMap,
  InferNextStateFromHandlerMap,
} from './create-handler-map'
import { merge } from './utils'
import { AnyAction } from './create-action'

/**
 * Reducer factory
 * @description combines multiple handler map into single reducer
 * @example
 * const counter = createReducer(0, handleAction => [
 *   handleAction(increment, state => state + 1),
 *   handleAction(decrement, state => state - 1),
 * ])
 */
export function createReducer<
  TPrevState,
  THandlerMap extends HandlerMap<TPrevState, any, any>
>(
  defaultState: TPrevState,
  handlerMapsCreator: (handle: CreateHandlerMap<TPrevState>) => THandlerMap[]
) {
  const handlerMap = merge(...handlerMapsCreator(createHandlerMap))

  return (
    state = defaultState,
    action: InferActionFromHandlerMap<THandlerMap> | AnyAction
  ): InferNextStateFromHandlerMap<THandlerMap> => {
    const handler = handlerMap[(<any>action).type]

    return handler ? handler(<any>state, action) :
      handlerMap.default ? handlerMap.default(<any>state, action) :
        state
  }
}
