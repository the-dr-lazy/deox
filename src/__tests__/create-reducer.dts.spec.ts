import { createReducer } from '../create-reducer'
import { createActionCreator } from '../create-action-creator'
import { DeepImmutable } from '../types'

// @dts-jest:group createReducer
{
  const increment = createActionCreator('INCREMENT')
  const decrement = createActionCreator('DECREMENT')
  const reset = createActionCreator('RESET', resolve => (count: number) =>
    resolve(count)
  )

  // @dts-jest:pass:snap counter scenario
  createReducer(0, handleAction => [
    handleAction(increment, state => state + 1),
    handleAction(decrement, state => state + 1),
    handleAction(reset, (_state, { payload }) => payload),
  ])
}

{
  type Item = { name: string }
  type ItemsState = Item[]

  const identity = createActionCreator('IDENTITY')
  const addItem = createActionCreator('ADD_ITEM', resolve => (name: string) =>
    resolve(name)
  )
  const removeItem = createActionCreator(
    'REMOVE_ITEM',
    resolve => (name: string) => resolve(name)
  )

  // @dts-jest:pass:snap mutable items list scenario
  createReducer(<ItemsState>[], handleAction => [
    handleAction(identity, state => state),
    handleAction(addItem, (state, { payload }) => [
      ...state,
      { name: payload },
    ]),
    handleAction(removeItem, (state, { payload }) =>
      state.filter(({ name }) => name !== payload)
    ),
  ])
}

{
  type Item = { name: string }
  type ItemsState = DeepImmutable<Item[]>

  const identity = createActionCreator('IDENTITY')
  const addItem = createActionCreator('ADD_ITEM', resolve => (name: string) =>
    resolve(name)
  )
  const removeItem = createActionCreator(
    'REMOVE_ITEM',
    resolve => (name: string) => resolve(name)
  )

  // @dts-jest:pass:snap immutable items list scenario
  createReducer(<ItemsState>[], handleAction => [
    handleAction(identity, state => state),
    handleAction(addItem, (state, { payload }) => [
      ...state,
      { name: payload },
    ]),
    handleAction(removeItem, (state, { payload }) =>
      state.filter(({ name }) => name !== payload)
    ),
  ])
}

{
  const increment = createActionCreator('INCREMENT')

  // @dts-jest:pass:snap counter scenario
  createReducer(0, handleAction => [
    handleAction(increment, state => state + 1),
    handleAction.others(() => 0),
  ])
}
