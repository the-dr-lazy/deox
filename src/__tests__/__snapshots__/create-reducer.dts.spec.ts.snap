// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createReducer counter scenario (type) should match snapshot 1`] = `"(state: number | undefined, action: { type: string; } | { type: \\"INCREMENT\\"; } | { type: \\"DECREMENT\\"; } | { type: \\"RESET\\"; payload: number; }) => number"`;

exports[`createReducer counter scenario (type) should match snapshot 2`] = `"(state: number | undefined, action: { type: string; } | { type: \\"INCREMENT\\"; }) => number"`;

exports[`createReducer immutable items list scenario (type) should match snapshot 1`] = `"(state: DeepImmutableArray<Item> | undefined, action: { type: string; } | { type: \\"IDENTITY\\"; } | { type: \\"ADD_ITEM\\"; payload: string; } | { type: \\"REMOVE_ITEM\\"; payload: string; }) => DeepImmutableArray<Item> | DeepImmutableObject<Item>[]"`;

exports[`createReducer mutable items list scenario (type) should match snapshot 1`] = `"(state: Item[] | undefined, action: { type: string; } | { type: \\"IDENTITY\\"; } | { type: \\"ADD_ITEM\\"; payload: string; } | { type: \\"REMOVE_ITEM\\"; payload: string; }) => Item[]"`;
